<a href="https://colab.research.google.com/github/zeph0yr/Operation-Research-Optimization/blob/master/Lab10.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!pip install pulp


     |████████████████████████████████| 40.6MB 110kB/s 


In [ ]:
from pulp import *
import numpy as np
import pandas as pd

#Question 1
##Model

Minimize z=
$120*x_1 + 100*x_2 + 80*x_3 + 80*x_4 + 90*x_5 + 110*x_6 + 110*x_7 + 140*x_8 +120*x_9 $

###Subject To

$x_1 + x_2 + x_3 = 1$

$x_4 + x_5 + x_6 = 1$

$x_7 + x_8 + x_9 = 1$

$x_1 + x_4 + x_7 = 1$

$x_2 + x_5 + x_8 = 1$

$x_3 + x_6 + x_9 = 1$



##Upload .csv file

In [ ]:
from google.colab import files

data=files.upload()

for x in data.keys():
  print("User uploaded file {} with size = {} bytes".format(x,len(data[x])))

Saving LAB10_Q1.csv to LAB10_Q1.csv
User uploaded file LAB10_Q1.csv with size = 70 bytes


##Print .csv file

In [ ]:
df=pd.read_csv('LAB10_Q1.csv')
df

A    B    C  Availability
0  120  100   80           1.0
1   80   90  110           1.0
2  110  140  120           1.0
3    1    1    1           NaN

In [ ]:
nvar=9
ncons=3+3
coeff=df.iloc[0:3,0:3].to_numpy()
availability=df['Availability'].to_numpy()
demand=df.iloc[3:4,0:3].to_numpy()

##Define Linear programming problem

In [ ]:
model=LpProblem("LAB10_Q1.csv",LpMinimize)

##Define decision variables

In [ ]:
x={i:LpVariable(name=f"x{i}",lowBound=0,cat='Integer') for i in range(1,nvar+1)}

##Define objective function

In [ ]:
model+=lpSum(coeff[i,j]*x[(i*3) + j+1]for i in range(0,3) for j in range(0,3))

##Add constraints (row-wise)

In [ ]:
for row in range(0,3):
  model+=lpSum(x[(row*3)+column+1] for column in range(0,3))==availability[row]

##Add Constraints (column-wise)

In [ ]:
for column in range(0,3):
  model+=lpSum(x[(row*3)+column+1] for row in range(0,3))==demand[0][column]

##Print model

In [ ]:
print(model)

LAB10_Q1.csv:
MINIMIZE
120*x1 + 100*x2 + 80*x3 + 80*x4 + 90*x5 + 110*x6 + 110*x7 + 140*x8 + 120*x9 + 0
SUBJECT TO
_C1: x1 + x2 + x3 = 1

_C2: x4 + x5 + x6 = 1

_C3: x7 + x8 + x9 = 1

_C4: x1 + x4 + x7 = 1

_C5: x2 + x5 + x8 = 1

_C6: x3 + x6 + x9 = 1

VARIABLES
0 <= x1 Integer
0 <= x2 Integer
0 <= x3 Integer
0 <= x4 Integer
0 <= x5 Integer
0 <= x6 Integer
0 <= x7 Integer
0 <= x8 Integer
0 <= x9 Integer



##Get the results

In [ ]:
status=model.solve()
print(f"status: {model.status},{LpStatus[model.status]}")
print(f"Objective= {model.objective.value()}")

for var in x.values():
  print("{}={}".format(var.name,var.value()))

status: 1,Optimal
Objective= 280.0
x1=0.0
x2=0.0
x3=1.0
x4=0.0
x5=1.0
x6=0.0
x7=1.0
x8=0.0
x9=0.0


#Question 2
##Model

###Minimize z=

$10*x_1 + 6*x_{10} + 10*x_{11} + 7*x{12} + 2*x_{13} + 2*x_{14} + 2*x_{15} + $ $ 7*x_{16} + 11*x_{17} + 9*x_{18} + 7*x_{19} + 5*x_{2} + 12*x_{20} + 7*x_{21}$ + $9*x_{22} + 10*x_{23} + 4*x_{24} + 12*x_{25} + 13*x_3 + 15*x_4 + 16*x_5 + 3*x_6$ +$ 9*x_7 + 18*x_8 + 13*x_9 $

###Subject to:
$x_1 + x_2 + x_3 + x_4 + x_5 = 1$

$x_{10} + x_6 + x_7 + x_8 + x_9 = 1$

$x_{11} + x_{12} + x_{13} + x_{14} + x_{15} = 1$

$x{16} + x_{17} + x_{18} + x_{19} + x_{20} = 1$

$x_{21} + x_{22} + x_{23} + x_{24} + x_{25} = 1$

$x1 + x_{11} + x_{16} + x_{21} + x_6 = 1$

$x_{12} + x_{17} + x_2 + x_{22} + x_7 = 1$

$x_{13} + x_{18} + x_{23} + x_3 + x_8 = 1$

$x_{14} + x_{19} + x_{24} + x_4 + x_9 = 1$

$x_{10} + x_{15} + x_{20} + x_{25} + x_5 = 1$

##Upload .csv file

In [ ]:
from google.colab import files
data=files.upload()
for x in data.keys():
  print(f"User uploaded file {x} with size = {len(data[x])} bytes")

Saving LAB10_Q2.csv to LAB10_Q2 (1).csv
User uploaded file LAB10_Q2.csv with size = 112 bytes


##Print .csv file 

In [ ]:
df=pd.read_csv('LAB10_Q2.csv')
df

A   B   C   D   E  Availability
0  10   5  13  15  16           1.0
1   3   9  18  13   6           1.0
2  10   7   2   2   2           1.0
3   7  11   9   7  12           1.0
4   7   9  10   4  12           1.0
5   1   1   1   1   1           NaN

In [ ]:
nvar=5*5
ncons=5+5
coeff=df.iloc[0:5,0:5].to_numpy()
availability=df['Availability'].to_numpy()
demand=df.iloc[5:6,0:5].to_numpy()

##Define Linear programming problem

In [ ]:
model=LpProblem("LAB10_Q2",LpMinimize)

##Define Decision variables

In [ ]:
x={i:LpVariable(name=f"x{i}",lowBound=0,cat='Integer') for i in range(1,nvar+1)}

##Define Objective Function

In [ ]:
model+=lpSum(coeff[row][column]*x[(row*5)+column+1] for row in range(0,5) for column in range(0,5))

##Add constraints (row-wise)

In [ ]:
for row in range(0,5):
  model+=lpSum(x[(row*5)+column+1] for column in range(0,5))==availability[row]
  

##Add Constraints (Column-wise)

In [ ]:
for column in range(0,5):
  model+=lpSum(x[(row*5)+column+1] for row in range(0,5))==demand[0][column]

##Print Model

In [ ]:
print(model)

LAB10_Q2:
MINIMIZE
10*x1 + 6*x10 + 10*x11 + 7*x12 + 2*x13 + 2*x14 + 2*x15 + 7*x16 + 11*x17 + 9*x18 + 7*x19 + 5*x2 + 12*x20 + 7*x21 + 9*x22 + 10*x23 + 4*x24 + 12*x25 + 13*x3 + 15*x4 + 16*x5 + 3*x6 + 9*x7 + 18*x8 + 13*x9 + 0
SUBJECT TO
_C1: x1 + x2 + x3 + x4 + x5 = 1

_C2: x10 + x6 + x7 + x8 + x9 = 1

_C3: x11 + x12 + x13 + x14 + x15 = 1

_C4: x16 + x17 + x18 + x19 + x20 = 1

_C5: x21 + x22 + x23 + x24 + x25 = 1

_C6: x1 + x11 + x16 + x21 + x6 = 1

_C7: x12 + x17 + x2 + x22 + x7 = 1

_C8: x13 + x18 + x23 + x3 + x8 = 1

_C9: x14 + x19 + x24 + x4 + x9 = 1

_C10: x10 + x15 + x20 + x25 + x5 = 1

VARIABLES
0 <= x1 Integer
0 <= x10 Integer
0 <= x11 Integer
0 <= x12 Integer
0 <= x13 Integer
0 <= x14 Integer
0 <= x15 Integer
0 <= x16 Integer
0 <= x17 Integer
0 <= x18 Integer
0 <= x19 Integer
0 <= x2 Integer
0 <= x20 Integer
0 <= x21 Integer
0 <= x22 Integer
0 <= x23 Integer
0 <= x24 Integer
0 <= x25 Integer
0 <= x3 Integer
0 <= x4 Integer
0 <= x5 Integer
0 <= x6 Integer
0 <= x7 Integer
0 <= x8 I

##Get the results

In [ ]:
status=model.solve()
print(f"status= {model.status}, {LpStatus[model.status]}")
print(f"Objective={model.objective.value()}")

for var in x.values():
  print(f"{var.name}={var.value()}")

status= 1, Optimal
Objective=23.0
x1=0.0
x2=1.0
x3=0.0
x4=0.0
x5=0.0
x6=1.0
x7=0.0
x8=0.0
x9=0.0
x10=0.0
x11=0.0
x12=0.0
x13=0.0
x14=0.0
x15=1.0
x16=0.0
x17=0.0
x18=1.0
x19=0.0
x20=0.0
x21=0.0
x22=0.0
x23=0.0
x24=1.0
x25=0.0
